Task/question

macfrei opened the pull request:
Added some basic styling, a function which determines the result of the house quiz and so on. Please feel free to comment as much as you like. I have the feeling that the solution could be more elegant... 👼. Tried to dabble with A11Y a bit, that's why you see `scrollToTop` and `getElementById`: https://andyfry.co/multi-step-form-accessibility/

A start screen and a result pop up will follow. For now, only console logs xD

macfrei committed:
Install a11y addon for storybook

macfrei committed:
Update props of Question story

macfrei committed:
Add focus to labels

macfrei committed:
Change ternary to conditional rendering

macfrei committed:
Remove console logs

macfrei committed:
Add font and some global styles

macfrei committed:
Add basic page layout

macfrei committed:
Add basic form styles

macfrei committed:
Refine question styles

macfrei committed:
Update button text and tests

macfrei committed:
Add advanceStep hook with a11y features

macfrei committed:
Add functionality to find house with most corresponding answers

macfrei committed:
Write simple test for find house function

macfrei committed:
Move functionality to its own file

mbosselmann commented on public/index.html:
```diff
@@ -23,6 +23,11 @@
+    <link
+      href="https://fonts.googleapis.com/css2?family=Galdeano&display=swap"
+      rel="stylesheet"
+    />
```
I would recommend you to upload fonts and not use the google link option anymore. [See this article to learn more about the why.](https://www.heise.de/news/DSGVO-Abmahnwelle-wegen-Google-Fonts-7206364.html) :)

rolfkoizumi replied:
Just a FYI (you might know this already): Linking google fonts directly from Googles servers is easy, but as far as I know German data protection laws (DSGVO) would require you to ask the users consent before fetching these fonts, or at least mention their usage on your privacy page.
So if you at some point create a page for a German/European audience and want to be on the safe side, it might be easier to just download the fonts from Google and host them yourself.

mbosselmann commented on src/App.js:
```diff
@@ -1,12 +1,39 @@
   function saveHouse(data) {
-    console.log(data)
+    const house = findHouse(data, questions)
+    console.log(house)
```
The console.log() is not necessary anymore or? ;)

mbosselmann replied:
Ah now I see why there is a console.log. 🙈

mbosselmann commented on src/App.js:
```diff
@@ -1,12 +1,39 @@
   }
 }
 
+const PageLayout = styled.div`
```
Instead of using a `div`, you could also work with a `main` element as this is more semantic or? 🤔

macfrei replied:
Good idea!

mbosselmann commented on src/components/Form.js:
```diff
@@ -17,21 +18,21 @@ Form.propTypes = {
-      ) : (
-        <button>Submit</button>
       )}
+      {currentStep === totalSteps && <button>Reveal Hogwarts house!</button>}
```
The button type needs to be added. :)

mbosselmann commented on src/components/Form.js:
```diff
@@ -17,21 +18,21 @@ Form.propTypes = {
   const { question, options } = questions[currentStep]
 
   return (
     <FormStyled onSubmit={handleSubmit}>
```
[According to this article on MDN ](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/form_role#description)a form needs an accessible name to be recognized as a landmark. Maybe you can add a heading and reference to it with `aria-labelledby` or you can use `aria-label`.

macfrei replied:
I wanted to add a bit more description later on in form of a landing page. But will add an `aria-label` 👍

mbosselmann commented on src/components/Form.js:
```diff
@@ -47,13 +48,24 @@ export default function Form({ questions, onSaveHouse }) {
   max-width: 800px;
   padding: 20px;
+
+  button {
```
I would prefer to put these stylings in a styled button component. :)

mbosselmann commented on src/components/Form.spec.js:
```diff
@@ -20,7 +20,7 @@ describe('Form', () => {
+    expect(screen.getByRole('button', { name: /reveal/i })).toBeInTheDocument()
   })
 
   it('should render a button with text "submit" when at last question', () => {
```
Your test description needs to be updated as the button text is now reveal? :)

mbosselmann commented on src/components/Form.spec.js:
```diff
@@ -72,7 +72,7 @@ describe('Form', () => {
     userEvent.click(nextButton)
```
Can I ask why you do not initialize an user event with `const user = userEvent.setup()`first? :) [See Testing Library docs for user event setup.](https://testing-library.com/docs/user-event/setup)

macfrei replied:
I started this project long before the `userEvent` changed and when starting again, I didn't check the tests 😆

mbosselmann commented on src/hooks/useAdvanceStep.js:
```diff
@@ -0,0 +1,30 @@
+  const changeTitle = title => (document.title = title)
+  const setFocus = () => {
+    setTimeout(() => {
+      const focusTarget = document.getElementById('focus-target')
```
Instead of working with `getElementById` I would prefer to use `querySelector` in combination with a `data` attribute. But that is up to you to decide. :)

mbosselmann commented on src/services/findHouse.js:
```diff
@@ -0,0 +1,41 @@
+  const listOfHousesAccordingToAnswers = houseQuizAnswers.flatMap(
+    answerArray => {
+      const associatedQuestion = questionArray.find(
+        el => el.question === answerArray[0]
```
I would write out all terms and not abbreviate.

mbosselmann commented on src/services/findHouse.js:
```diff
@@ -0,0 +1,41 @@
+      const houses = possibleAnswers
+        .filter(answer => {
+          // Because answer could be an array of multiple possibilities
+          if (answer.filter(e => e.includes(answerArray[1])).length > 0) {
```
Same here: I would write out all terms and not abbreviate.

mbosselmann reviewed (commented):
Hi Lene,
you rock! 🪨 
I left some remarks for you to consider. Feel tree to ask me if you have further questions. :)
Looking forward to seeing more of your project.

rolfkoizumi commented on src/App.js:
```diff
@@ -1,12 +1,39 @@
   function saveHouse(data) {
-    console.log(data)
+    const house = findHouse(data, questions)
+    console.log(house)
```
Is this still work in progress? Or does `findHouse()` do more than returning something?

macfrei replied:
This is still work in progress, yes :)

And `findHouse()` does just return a value.

rolfkoizumi commented on src/GlobalStyles.js:
```diff
@@ -1,13 +1,35 @@
+}
+
 body {
+    margin:0;
```
Missing space: `margin: 0;`

rolfkoizumi commented on src/components/Question.js:
```diff
@@ -9,7 +9,7 @@ Question.propTypes = {
 export default function Question({ question, options, onChange }) {
   return (
-    <QuestionStyled>
+    <QuestionStyled id="focus-target">
```
I'd recommend not to use `id` as selector for either css nor JS-functionality. I think nowadays `id` should mainly be used semantically as anchor (like in https://mysite.com/some-page#specific-section).

If you're using it as an anchor here, I would give it a more semantic name, since it'll be shown to the user in the url.

If you're only using it as selector (and I think you are), I would rather use an data-attribute, like this: `<QuestionStyled data-focus-target>`
And then you can grab it with: `document.querySelector('[data-focus-target]')` (or `querySelectorAll` if there could be multiple elements).

The reason is that when changing/refactoring stuff, developers often overlook that an id is used as a selector, especially when the corresponding JS is locating in another file. In this case, the naming "focus-target" implies that this id is a target, which helps, but I think it's still better to have explicit attributes for different usages.

rolfkoizumi commented on src/services/findHouse.js:
```diff
@@ -0,0 +1,41 @@
+  ).sort((a, b) => b[1] - a[1])[0][0]
+
+  return house
+}
```
This function looks overly complicated (with a lot of nested loops inside) and I have the feeling that the data structure is also not optimized.

Here is an example how you could refactor this current function which I think might be a bit more readable:
```
function getHighestOccurrence(array) {
  return Object.entries(
    array.reduce((occurences, element) => {
      if (element in occurences) {
        occurences[element]++
      } else {
        occurences[element] = 1
      }
      return occurences
    }, {})
  ).sort((a, b) => b[1] - a[1])[0][0]
}

export default function calculateHouse(quizResults, questionArray) {
  if (!Object.keys(quizResults).length) {
    console.error('Something is wrong with your data, Muggle!')
    return undefined;
  }

  const houses = Object.entries(quizResults).flatMap(([question, answer]) => {
    const answers = questionArray.find(q => q.question === question).answers

    const housesPerAnswer = Object.entries(answers).reduce((housesPerAnswer, [house, houseAnswer]) => {
      housesPerAnswer[houseAnswer] = [
        ...(housesPerAnswer[houseAnswer] ?? []),
        house,
      ]

      return housesPerAnswer
    }, {})

    return housesPerAnswer[answer]
  })

  return getHighestOccurrence(houses)
}
```

And with some restructuring of the `quizResults` and `questionArray`, I think it can be made even simpler.

rolfkoizumi commented on src/services/findHouse.spec.js:
```diff
@@ -0,0 +1,46 @@
+import findHouse from './findHouse'
+
+const data = {
```
Why not call this variable `quizResults`, same as `findHouse()` expects?

And I'm not that much a fan of having an object with sentences as keys. What will happen if you use unicode in there (ö, ä)? Or any other symbols? Maybe a question will at some point contain html, images or something.

I think I would rather use an array for `quizResults` with each result being an object with the question-id and the answer. The questions even have ids already and it'd be much safer to find your question via its id than having to do comparisons of whole sentences.

rolfkoizumi commented on src/services/findHouse.spec.js:
```diff
@@ -0,0 +1,46 @@
+      Ravenclaw: 'Hunger',
+      Hufflepuff: 'Cold',
+      Slytherin: 'Boredom',
+    },
```
And I might restructure the answers to something like:
```
answers: {
    Dawn: ['Gryffindor', 'Ravenclaw'],
    Dusk: ['Hufflepuff'],
    Foo: ['Slytherin'],
},
```

If these are used only inside `findHouse`, I think this structure would make things easier.

miriamhaenle commented on src/components/Question.js:
```diff
@@ -27,12 +27,56 @@ export default function Question({ question, options, onChange }) {
-  gap: 16px;
+  display: grid;
+  place-items: center;
+  grid-template-columns: 1fr 1fr;
+  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
```
Doesn't the line 33 override 32?

macfrei replied:
Yes 🤣 forgot to delete it! Nice catch!

miriamhaenle commented on src/hooks/useAdvanceStep.js:
```diff
@@ -0,0 +1,30 @@
+  const changeTitle = title => (document.title = title)
+  const setFocus = () => {
+    setTimeout(() => {
+      const focusTarget = document.getElementById('focus-target')
```
Did you try passing a ref to the hook for this component instead of querying from the document? I always thought this is a anti pattern in react.

Maybre try to pass the form as ref here :)

macfrei replied:
I'll look into refs. Haven't really worked with them before and also felt a bit weird to do it like this as I don't like to call the `document` directly in React.

miriamhaenle commented on src/services/findHouse.js:
```diff
@@ -0,0 +1,41 @@
+      )
+      const possibleAnswers = Object.entries(associatedQuestion.answers) // [[house, answer], ...]
+
+      const houses = possibleAnswers
```
you could directly return the array as the additional variable doesn't add much to the readability (and it is clear from the function name that it will return houses)

macfrei committed:
Small changes according to code review

macfrei committed:
Use "ref" insetead of talking to DOM directly

macfrei committed:
Change data structure

macfrei committed:
Change form data structure

macfrei committed:
Remove unnecessary styling

macfrei committed:
Update and simplify findHouse function

macfrei committed:
Add disable attribute to button

macfrei committed:
Remove google fonts

macfrei commented:
Thank you all for your input 🙇‍♀️ . I have worked a bit on your feedback:

- Used `ref` instead of `document.get...`
- Changed data structure
- Simplified `findHouse` function
- Deleted Google fonts (for now)

Let me know what you think :)

jamarob commented on src/components/Form.js:
```diff
@@ -3,6 +3,7 @@ import styled from 'styled-components/macro'
 import Question from './Question'
+import useAdvanceStep from '../hooks/useAdvanceStep'
 
 Form.propTypes = {
```
use TypeScript ;)

jamarob commented on src/hooks/useAdvanceStep.js:
```diff
@@ -0,0 +1,34 @@
+    }, 0)
+  }
+
+  function advanceStep() {
```
why function?

jamarob commented on src/hooks/useAdvanceStep.js:
```diff
@@ -0,0 +1,34 @@
+
+  const [currentStep, setCurrentStep] = useState(0)
+
+  const scrollTop = () => window.scrollTo({ x: 0 })
+
+  const changeTitle = title => (document.title = title)
```
No local dependecies, so they can be moved outside of the component (and not get redefined on every update)

jamarob commented on src/hooks/useAdvanceStep.js:
```diff
@@ -0,0 +1,34 @@
+
+  const changeTitle = title => (document.title = title)
+
+  const setFocus = () => {
+    setTimeout(() => {
+      ref.current.setAttribute('tabindex', '-1')
+      ref.current.focus()
+      ref.current.removeAttribute('tabindex')
+    }, 0)
+  }
```
has dependecy on ref but since ref is constant can be wrapped in useCallback to prevent unnecessary redefinitions.

jamarob commented on src/services/findHouse.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+function getHighestOccurrence(houses) {
+  const house = Object.entries(
+    houses.reduce((acc, cur) => {
+      if (cur in acc) {
+        acc[cur]++
+      } else {
+        acc[cur] = 1
+      }
+      return acc
+    }, {})
+  ).sort((a, b) => b[1] - a[1])[0][0]
```
🤯 can you give a speaking name to the reducer (or say what acc and cur are) to explain what it is doing.

macfrei replied:
I will. Will merge this one though as it has gotten rather big already. Will note your suggestions to implement asap, thank you! 😊
